x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineFilters.java

```diff
@@ -143,28 +166,38 @@ private static LogicalPlan pushDownPastJoin(Filter filter, Join join, FoldContex
             LogicalPlan left = join.left();
             LogicalPlan right = join.right();
 
-            // split the filter condition in 3 parts:
-            // 1. filter scoped to the left
-            // 2. filter scoped to the right
-            // 3. filter that requires both sides to be evaluated
-            ScopedFilter scoped = scopeFilter(Predicates.splitAnd(filter.condition()), left, right);
+            // Split the filter condition in 3 parts.
+            // For InlineJoin we use a scoping that allows pushing down filters either to right side only or to both sides.
+            // For the rest of the joins we use the standard scoping:
+            // - filters scoped to the left
+            // - filters scoped to the right
+            // - filter that requires both sides to be evaluated
+            ScopedFilter scoped = join instanceof InlineJoin ij
+                ? scopeInlineStatsFilter(Predicates.splitAnd(filter.condition()), ij)
+                : scopeFilter(Predicates.splitAnd(filter.condition()), left, right);
+
+            // For InlineJoin, in EsqlSession the left-hand side becomes the source for the right-hand side replacing the StubRelation
+            // and, as such, the filters used on the left-hand side are also used on the right-hand side.
+            var pushableToLeftSide = join instanceof InlineJoin ? scoped.commonFilters() : scoped.leftFilters();
+            var pushableToRightSide = scoped.rightFilters();
+            var commonFilters = join instanceof InlineJoin ? scoped.leftFilters() : scoped.commonFilters();
             boolean optimizationApplied = false;
             // push the left scoped filter down to the left child
-            if (scoped.leftFilters.size() > 0) {
+            if (pushableToLeftSide.size() > 0) {
                 // push the filter down to the left child
-                left = new Filter(left.source(), left, Predicates.combineAnd(scoped.leftFilters));
+                left = new Filter(left.source(), left, Predicates.combineAnd(pushableToLeftSide));
                 // update the join with the new left child
                 join = (Join) join.replaceLeft(left);
                 // we completely applied the left filters, so we can remove them from the scoped filters
-                scoped = new ScopedFilter(scoped.commonFilters(), List.of(), scoped.rightFilters);
+                scoped = new ScopedFilter(commonFilters, List.of(), scoped.rightFilters);
```

Comment:
```suggestion
                scoped = new ScopedFilter(commonFilters, List.of(), pushableToRightSide);
```
Here and below: it's hard to track which filters come from where. The suggestion above is one way to go, but another would be to rewrite `scoped` above instead of extracting the three variables.

Reply:
I've tried multiple approaches here, and none of them is ideal. If I do it like you say (and I believe I did try this approach) another decision based on the join type must be taken somewhere further below. Because the logic is common in many places for inlinejoin and other types of join, I really wanted to not duplicate the logic. Initially I also used a single method for scoping, but it was too messy. Imho, the code I came up with is the least intrusive one and somewhat clean.

Likely, I didn't find the best approach; please, if you have the time, try to make it more easily understandable by refactoring the way you think best. From **my** point of view (and ease of grasping the logic), what is in this PR is the best version.